### PR TITLE
Modify Dispatch region creation to only fuse Matmuls with compatible ops.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The fusion of Matmul/BatchMatmul with element-wise operations only works
for consumers where the output shape is compatible with the output of the
op. Adapt the dispatch region creation logic to account for that.

Also make reshape always fused with its producers or consumers.

Note that these enhancements to dispatch region creation are mostly
temporary and just to tide over things as the underlying codegeneration
gets more features.

### CODE REVIEW

File: iree/test/e2e/regression/matmul_add.mlir [hunk 1]
```diff
@@ -1,26 +1,271 @@
...
+    %4 = "mhlo.dot"(%arg0, %arg1) :
+      (tensor<16x32xf32>, tensor<32x48xf32>) -> tensor<16x48xf32>
+    %5 = mhlo.add %4, %arg2 : tensor<16x48xf32>
+    flow.return %5 : tensor<16x48xf32>
+  }
+  return %3 : tensor<16x48xf32>
 }
+//      CHECK: EXEC @matmul_add_large
+//      CHECK: 16x48xf32=
+// CHECK-SAME: [196 196 196 196 196 196 196 196 196 196 196 196
```
Comment (benvanik) [inline iree/test/e2e/regression/matmul_add.mlir:52]: feels suspicious - is there something more useful we can use to ensure this is producing the correct results without transposition/etc?

Comment (MaheshRavishankar) [reply iree/test/e2e/regression/matmul_add.mlir:52]: The answer is right. Each element is (2.0 * 3.0) * 32 + 4.0 = 196

Comment (MaheshRavishankar) [reply iree/test/e2e/regression/matmul_add.mlir:52]: Tried to do something more robust (though I think what is there is fine). Tried to take the output of the following numpy program

```
import numpy as np

np.set_printoptions(threshold=8192, precision=6)
a = np.arange(0.0, 1.0, 1.0/(16*32), np.float32).reshape((16, 32))
b = np.arange(0.0, 1.0, 1.0/(32*48), np.float32).reshape((32, 48))
c = np.arange(0.0, 1.0, 1.0/(16*48), np.float32).reshape((16, 48))
d = np.matmul(a, b) + c
print (repr(a))
print (repr(b))
print (repr(c))
print (d)
```

But the `iree-run-mlir` doesnt have the check dialect. So testing the output becomes almost impossible

Comment (MaheshRavishankar) [reply iree/test/e2e/regression/matmul_add.mlir]: Moved the test to `structural/...` to  use the `check.*` dialect for correctness.

File: iree/compiler/Dialect/Flow/Transforms/DispatchConfig.cpp [hunk 2]
```diff
@@ -48,8 +48,26 @@ bool isIndexOp(Operation *op) {
...
+/// Returns true if |lhs| and |rhs| return a single value with the same shape.
+bool checkSameOutputShape(Operation *lhs, Operation *rhs) {
+  if (lhs->getNumResults() != 1 || rhs->getNumResults() != 1) {
+    return false;
+  }
+  ShapedType lhsType = lhs->getResults()[0].getType().dyn_cast<ShapedType>();
+  ShapedType rhsType = rhs->getResults()[0].getType().dyn_cast<ShapedType>();
+  if (!lhsType || !rhsType || lhsType != rhsType) {
+    return false;
+  }
```
Comment (hanhanW) [inline iree/compiler/Dialect/Flow/Transforms/DispatchConfig.cpp:63]: [optional] `if (!lhsType || lhsType != rhsType)` is enough, but I think maybe the current way is easier to understand?

**Review by hanhanW — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchConfig.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchConfig.cpp:63]: True. But as you said, this is easier to understand the condition,

**Pull request merged**